### PR TITLE
Improve the guide on Pull

### PR DESCRIPTION
**Pull request by zainab-ali**
This PR improves the section of the guide on `fs2.Pull`. Feedback is of course very welcome!

I've also upgraded the JDK version in the nix shell to 16. This is necessary for compiling the `java.net.UnixDomainSocketAddress` ([JEP 380](https://openjdk.java.net/jeps/380)).

### CODE REVIEW

File: shell.nix [hunk 1]
```diff
@@ -7,13 +7,6 @@ let
     packageOverrides = p: rec {
       sbt = p.sbt.overrideAttrs (
         old: rec {
```
Comment (zainab-ali) [inline shell.nix:9]: The sbt version in nixos-21.05 is `1.5.4`. I don't think we need to override it.

File: shell.nix [hunk 2]
```diff
@@ -23,9 +16,9 @@ let
   };
 
   nixpkgs = builtins.fetchTarball {
-    name   = "nixos-unstable-2021-01-03";
-    url    = "https://github.com/NixOS/nixpkgs/archive/56bb1b0f7a3.tar.gz";
-    sha256 = "1wl5yglgj3ajbf2j4dzgsxmgz7iqydfs514w73fs9a6x253wzjbs";
+    name   = "nixos-21.05";
```
Comment (zainab-ali) [inline shell.nix:19]: I've upgraded the channel to `21.05`. This contains openjdk16.

File: site/documentation.md [hunk 3]
```diff
@@ -30,6 +30,7 @@
 * [Inference Driven Design](https://mpilquist.github.io/blog/2018/07/04/fs2/), by [Michael Pilquist][mpilquist], describes some of the tradeoffs in designing the API and the code of FS2, used to work around some of the problems in the Scala compiler.
 * [Tips for working with FS2](https://underscore.io/blog/posts/2018/03/20/fs2.html), by [Pere Villega](https://github.com/pvillega),
 * [A streaming library with a superpower: FS2 and functional programming](https://medium.freecodecamp.org/a-streaming-library-with-a-superpower-fs2-and-functional-programming-6f602079f70a).
+* [No leftovers: Working with pulls in fs2](https://blog.kebab-ca.se/chapters/fs2/overview.html), by [Zainab Ali](https://github.com/zainab-ali).
```
Comment (zainab-ali) [inline site/documentation.md:33]: I hope it's ok to add this here too?

File: site/guide.md [hunk 4]
```diff
@@ -283,13 +283,54 @@ else Some(c => c.size match {
...
 
+#### Transforming streams using pulls
+
 Sometimes, `scanChunksOpt` isn't powerful enough to express the stream transformation. Regardless of how complex the job, the `fs2.Pull` type can usually express it.
 
-The `Pull[+F[_],+O,+R]` type represents a program that may pull values from one or more streams, write _output_ of type `O`, and return a _result_ of type `R`. It forms a monad in `R` and comes equipped with lots of other useful operations. See the
-[`Pull` class](https://github.com/functional-streams-for-scala/fs2/blob/series/1.0/core/shared/src/main/scala/fs2/Pull.scala)
+The `Pull[F[_],O,R]` type represents a program that may pull values from one or more streams, write _output_ of type `O`, and return a _result_ of type `R`. It forms a monad in `R` and comes equipped with lots of other useful operations. See the
```
Comment (zainab-ali) [inline site/guide.md:290]: I don't think that the variance is important here, but am happy to be corrected.

File: site/guide.md [hunk 5]
```diff
@@ -283,13 +283,54 @@ else Some(c => c.size match {
...
+[`Pull` class](https://github.com/functional-streams-for-scala/fs2/blob/main/core/shared/src/main/scala/fs2/Pull.scala)
```
Comment (zainab-ali) [inline site/guide.md:291]: Series 1.0 is quite old. I think it makes more sense to link to the latest version of the code.

**Comment by mpilquist** [comment]
Yes!! I've been looking forward to this PR! Will review later today but so happy to see this. :)

File: site/guide.md [hunk 6]
```diff
@@ -601,8 +660,8 @@ def rows[F[_]](h: CSVHandle)(implicit F: Async[F]): Stream[F,Row] = {
 }
 ```
 
-See [`Queue`](https://github.com/functional-streams-for-scala/fs2/blob/series/1.0/core/shared/src/main/scala/fs2/concurrent/Queue.scala)
-for more useful methods. Most concurrent queues in FS2 support tracking their size, which is handy for implementing size-based throttling of the producer.
+See [`Queue`](https://github.com/typelevel/cats-effect/blob/series/3.x/std/shared/src/main/scala/cats/effect/std/Queue.scala)
```
Comment (zainab-ali) [inline site/guide.md:663]: Given that `Queue` is now in cats effect, I've pointed the link to the cats effect codebase.

File: shell.nix (hunk 1 above)
Comment (rossabaker) [reply shell.nix:9]: I get `1.5.1`?   Either way, it gets the job done.

**Review by rossabaker — APPROVED** [review]
I'll leave the substantial review to the experts, but I'm here to bless the Nix as working. :smile:

**Pull request merged**